libevmasm/Assembly.cpp

```diff
@@ -679,13 +679,22 @@ LinkerObject const& Assembly::assemble() const
 		// Append an INVALID here to help tests find miscompilation.
 		ret.bytecode.push_back(static_cast<uint8_t>(Instruction::INVALID));
 
-	for (auto const& [subIdPath, bytecodeOffset]: subRef)
 	{
-		bytesRef r(ret.bytecode.data() + bytecodeOffset, bytesPerDataRef);
-		toBigEndian(ret.bytecode.size(), r);
-		ret.append(subAssemblyById(subIdPath)->assemble());
+		std::map<Assembly const*, size_t> subAssemblyOffsets;
+		for (auto const& [subIdPath, bytecodeOffset]: subRef)
+		{
+			Assembly const* subAssembly = subAssemblyById(subIdPath);
+			bytesRef r(ret.bytecode.data() + bytecodeOffset, bytesPerDataRef);
+			if (size_t* subAssemblyOffset = util::valueOrNullptr(subAssemblyOffsets, subAssembly))
+				toBigEndian(*subAssemblyOffset, r);
+			else
+			{
+				toBigEndian(ret.bytecode.size(), r);
```

Comment:
I guess some assert like:
```
		assertThrow(numberEncodingSize(ret.bytecode.size()) <= bytesPerDataRef, AssemblyException, "");
```

Wouldn't hurt here either.
Might even technically be possible to violate this and get a truncated offset.

Reply:
Related:
semanticTests/various/code_access_content.sol
semanticTests/various/code_access_padding.sol
semanticTests/various/code_access_runtime.sol

are all cases in which there are references to subassemblies which don't occur in ``m_subs``. For such cases, it may also be advisable to check if everything is sound wrt the size estimates and potentially underestimating the required tag widths.